File: MeArm-3D/tests/sim2sim/mearmV1Baseline.py
# -*- coding: utf-8 -*-
"""MeArm-V1 黄金基线（`robot-package/mearm-v1/tests/cases/*.json`）的 **Python 只读门面**。

与 `frontend/tests/helpers/mearmV1Baseline.ts` 一一对应。两条验收链
（前端 vitest / Python pytest）必须读**同一批文件** —— 否则「抽象前后行为一致」
会被拆成两套互不相干的标准，而两边各自都能自己绿。

本模块**只读**、不产生任何数值：期望值一律来自 `tools/gen_mearm_v1_baseline.py`
实跑采集的落盘文件（spec §6 / §16）。

目录**不在这里拼**：取自 `robot-package/mearm-v1/manifest.yaml` 的 `tests.cases`
（经 Core 的 `robopkg.declared_path`）。Phase 2 搬迁时正因如此只需动一处。
"""
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]           # tests/sim2sim -> tests -> MeArm-3D
_CORE_PY = ROOT / "core" / "python"
if str(_CORE_PY) not in sys.path:
    sys.path.insert(0, str(_CORE_PY))

from robopkg import declared_path                    # noqa: E402

#: 本门面**只**服务 MeArm-V1：显式写 id，不读选择器 `default`
#: （否则改 default 会让黄金数据静默换成另一台机器人的快照）。
MEARM_V1 = "mearm-v1"


def baseline_dir() -> Path:
    """MeArm-V1 的用例目录 = 包在 manifest 里声明的 `tests.cases`。"""
    return declared_path(MEARM_V1, "tests.cases")

#: 冻结时使用的固定 seed（可复现性的锚点；换它等于换了一整套数据）
SEED = 20260914

FILES: dict[str, str] = {
    "joint": "joint_cases.json",
    "fk": "fk_cases.json",
    "ik": "ik_cases.json",
    "workspace": "workspace_cases.json",
}

_REQUIRED_META = ("model", "modelVersion", "robotId", "seed", "generator", "cases")


def load_doc(name: str) -> dict[str, Any]:
    """读取一份基线文件（含 `model` / `modelVersion` / `seed` 等元信息）。"""
    if name not in FILES:
        raise KeyError(f"未知基线文件 {name!r}（可用：{sorted(FILES)}）")
    path = baseline_dir() / FILES[name]
    if not path.is_file():
        raise FileNotFoundError(
            f"缺少黄金基线 {path}\n"
            "  生成：python robot-package/mearm-v1/tools/gen_mearm_v1_baseline.py\n"
            "  （需要同时具备 node 与 mujoco）"
        )
    doc = json.loads(path.read_text(encoding="utf-8"))
    missing = [k for k in _REQUIRED_META if k not in doc]
    if missing:
        raise ValueError(f"{path.name} 缺少字段 {missing}（文件被改动过？）")
    return doc


def load_cases(name: str) -> list[dict[str, Any]]:
    """只取用例数组。"""
    return list(load_doc(name)["cases"])


def tolerance(doc: dict[str, Any], key: str, default: float) -> float:
    """读基线**自带**的容差。

    刻意默认取文件里的值：容差是冻结契约的一部分，散落在测试文件里迟早与
    基线漂移。文件缺该键时用调用方给的保守缺省 —— 方向**只能是收紧**。
    """
    return float(doc.get("tolerances", {}).get(key, default))

File: MeArm-3D/frontend/tests/helpers/mearmV1Baseline.ts
/**
 * MeArm-V1 黄金基线（Golden Test Dataset）的读取助手。
 *
 * 数据由 `tools/gen_mearm_v1_baseline.py` 采集，落在
 * `robot-package/mearm-v1/tests/cases/*.json`（Phase 2 起随包），
 * **前后端 + Python 三方共用同一份文件** —— 目录由该包的 manifest 声明，见
 * `helpers/robotPackage.ts`（TS 侧唯一解析处）/ `robopkg.declared_path`（Python 侧）。
 *
 * ⚠️ 这里**只读**，不生成、不修补。生成入口只有一个：
 * `tools/gen_mearm_v1_baseline.py`。手改这些 JSON 等于伪造基线。
 */
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { declaredPath } from './robotPackage';

/** 本门面**只**服务 MeArm-V1：显式写 id，不读选择器 `default`。 */
const MEARM_V1 = 'mearm-v1';

/** 用例目录 —— 取自 `robot-package/mearm-v1/manifest.yaml` 的 `tests.cases` */
export const BASELINE_DIR = declaredPath(MEARM_V1, 'tests.cases');

/** 所有基线文件共有的头信息 */
export interface BaselineMeta {
  model: string;
  modelVersion: string;
  robotId: string;
  robotName: string;
  seed: number;
  generator: string;
  note: string;
}

export interface JointCase {
  id: string;
  label: string;
  joints: Record<string, number>;
}

export interface FkCase extends JointCase {
  tcpFrontend: number[];
  tcpFrontendRotation: number[];
  framesFrontend: Record<string, number[]>;
  tcpRef: number[];
  tcpMujoco: number[];
  framesRef: Record<string, number[]>;
}

export interface IkExpectation {
  success: boolean;
  branch?: string;
  joints?: Record<string, number>;
  residual?: number;
  azimuth?: number;
  relativeAngle?: number;
  tcpFrontend?: number[];
  tcpMujoco?: number[];
  reason?: string;
  joint?: string | null;
  message?: string;
  candidateCount?: number;
}

export interface IkCase {
  id: string;
  label: string;
  kind: string;
  target: number[];
  expect: IkExpectation;
}

export interface WorkspaceCase {
  id: string;
  label: string;
  kind: string;
  target: number[];
  sagittalDistance: number;
  reach: number[];
  geometry: {
    pivotZ: number;
    pivotR: number;
    l1: number;
    l2: number;
    toolOffset: number[];
  };
  expect: { success: boolean; reason: string | null };
}

function read<T>(file: string): T {
  const full = path.join(BASELINE_DIR, file);
  return JSON.parse(readFileSync(full, 'utf8')) as T;
}

export function loadJointCases() {
  return read<BaselineMeta & { count: number; cases: JointCase[] }>('joint_cases.json');
}
export function loadFkCases() {
  return read<BaselineMeta & { tolerances: Record<string, number>; cases: FkCase[] }>(
    'fk_cases.json',
  );
}
export function loadIkCases() {
  return read<BaselineMeta & { tolerances: Record<string, number>; cases: IkCase[] }>(
    'ik_cases.json',
  );
}
export function loadWorkspaceCases() {
  return read<BaselineMeta & { cases: WorkspaceCase[] }>('workspace_cases.json');
}

/**
 * 逐位一致断言（`Object.is` 语义）。
 *
 * 为什么不用 `toBeCloseTo`：本阶段唯一的验收原则是「抽象前后**逐位**一致」
 * （spec §19）。用一个"接近"的判据去验一件声称"完全相同"的事，
 * 恰好会把要找的那类缺陷放过去（差 1e-10 的耦合项写错就是这么漏掉的）。
 */
export function expectBitIdentical(actual: readonly number[], expected: readonly number[]): void {
  if (actual.length !== expected.length) {
    throw new Error(`长度不同：${actual.length} vs ${expected.length}`);
  }
  for (let i = 0; i < actual.length; i += 1) {
    if (!Object.is(actual[i], expected[i])) {
      throw new Error(`第 ${i} 位不同：${actual[i]} vs ${expected[i]}`);
    }
  }
}
